File: backend/Model/PurchaseOrder.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace YLWorks.Model
{
    public class PurchaseOrder: BaseEntity
    {
        public Guid Id { get; set; }
        public string PurchaseOrderNo { get; set; } = string.Empty;
        public Guid? FromCompanyId { get; set; }
        public Company? FromCompany { get; set; }

        public DateTime? PODate { get; set; }
        public DateTime? POReceivedDate { get; set; }

        public Guid? SupplierId { get; set; }
        public Company? Supplier { get; set; }

        public Guid? ClientId { get; set; }
        public Company? Client { get; set; }

        public string? Terms { get; set; }
        public Guid? ProjectId { get; set; }
        public Project? Project { get; set; }

        public Guid? QuotationId { get; set; }
        public Quotation? Quotation { get; set; }

        public int? TotalQuantity { get; set; }
        public decimal? Gross { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalAmount { get; set; }
        public string? TotalInWords { get; set; }

        public string? Remarks { get; set; }
        public string? Notes { get; set; }

        public Guid? PurchaseOrderId { get; set; }
        public Guid? SalesOrderId { get; set; }

        public string Status { get; set; } = "Draft"; // Draft, Sent, Accepted, PartiallyReceived, Completed, Cancelled

        public decimal? InvoicedAmount { get; set; }
        public string? InvoiceStatus { get; set; } = "NotInvoiced";
        // NotInvoiced | PartiallyInvoiced | FullyInvoiced

        public string? TermsAndCondition { get; set; }
        public string? BankDetails { get; set; }

        public string? Attachment { get; set; }
        public Guid? CreatedById { get; set; }
        public ICollection<PurchaseOrderStatusHistory> PurchaseOrderStatusHistories { get; set; } = new List<PurchaseOrderStatusHistory>();
        public ICollection<PurchaseOrderItem>? PurchaseOrderItems { get; set; } = new List<PurchaseOrderItem>();
        public ICollection<Invoice> Invoices { get; set; } = new List<Invoice>();

    }

    public class PurchaseOrderStatusHistory
    {
        public Guid Id { get; set; }
        public Guid PurchaseOrderId { get; set; }
        [JsonIgnore]
        public PurchaseOrder PurchaseOrder { get; set; } = null!;
        public string Status { get; set; } = string.Empty;
        public DateTime ActionAt { get; set; }
        public Guid? ActionUserId { get; set; }
        public User? ActionUser { get; set; }
        public string? Remarks { get; set; }
        public string? SignatureImage { get; set; } 
    }

    public class PurchaseOrderItem : BaseEntity
    {
        public Guid Id { get; set; }
        public Guid PurchaseOrderId { get; set; }
        public PurchaseOrder PurchaseOrder { get; set; } = null!;
        public string? Item { get; set; }
        public string? Description { get; set; }
        public decimal Quantity { get; set; }
        public string Unit { get; set; } = "Nos";
        public decimal UnitPrice { get; set; }
        public decimal Discount { get; set; }
        public decimal TotalPrice { get; set; }
        public decimal? ReceivedQuantity { get; set; } = 0;
    }

    public class POItemDto
    {
        public Guid Id { get; set; }
        public string? Item { get; set; }
        public string Description { get; set; }
        public string? Unit { get; set; }
        public decimal Quantity { get; set; }
        public decimal UnitPrice { get; set; }
        public decimal Discount { get; set; }
        public decimal TotalPrice { get; set; }
        public decimal? ReceivedQuantity { get; set; } = 0;
    }

    public class POItemBase
    {
        public Guid? Id { get; set; }
        public Guid? PurchaseOrderId { get; set; }
        public string? Item { get; set; }
        public string? Description { get; set; } = string.Empty;
        public decimal Quantity { get; set; }
        public string Unit { get; set; } = "Unit";
        public decimal UnitPrice { get; set; }
        public decimal Discount { get; set; }
        public decimal TotalPrice { get; set; }
    }

    public class POItemRequest : POItemBase { }

    public class UpdatePOItemRequest : POItemBase
    {
        public Guid? Id { get; set; }
    }


    public class CreatePORequest
    {
        public string PurchaseOrderNo { get; set; } = string.Empty;
        public Guid? FromCompanyId { get; set; }
        public DateTime? PODate { get; set; }
        public DateTime? POReceivedDate {  get; set; }
        public Guid? SupplierId { get; set; }
        public Guid? ClientId {  get; set; }
        public string? Terms { get; set; }

        public Guid? QuotationId { get; set; }
        public Guid? ProjectId { get; set; }

        public Guid? PurchaseOrderId { get; set; }
        public Guid? SalesOrderId {  get; set; }

        public decimal? Gross { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalAmount { get; set; }

        public string? Notes { get; set; }
        public string? Remarks { get; set; }
        public string? TermsAndConditions { get; set; }
        public string? BankDetails { get; set; }
        public int? TotalQuantity { get; set; }
        public IFormFile? Attachment { get; set; }

        public List<POItemRequest> PurchaseOrderItems { get; set; } = new();
    }

    public class UpdatePORequest
    {
        public Guid Id { get; set; } 
        public string PurchaseOrderNo { get; set; } = string.Empty;
        public Guid? FromCompanyId { get; set; }
        public DateTime? PODate { get; set; }
        public DateTime? POReceivedDate { get; set; }
        public Guid? SupplierId { get; set; }
        public Guid? ClientId { get; set; }
        public string? Terms { get; set; }
        public Guid? QuotationId { get; set; }
        public Guid? ProjectId { get; set; }
        public Guid? PurchaseOrderId {  get; set; }
        public Guid? SalesOrderId {  get; set; }
        public decimal? Gross { get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalAmount { get; set; }
        public string? Notes { get; set; }
        public string? Remarks { get; set; }
        public string? TermsAndConditions { get; set; }
        public string? BankDetails { get; set; }
        public int? TotalQuantity { get; set; }
        public IFormFile? Attachment { get; set; }

        public List<UpdatePOItemRequest> PurchaseOrderItems { get; set; } = new();
    }

    public class UpdatePOStatusRequest
    {
        public Guid Id { get; set; }         
        public string? Status { get; set; } 
        public string? Remarks { get; set; }  
        public string? SignatureImage { get; set; } 
    }

    public class DropdownResponseDto
    {
        public List<Company> Companies { get; set; } = new();
        public List<Company> Clients { get; set; } = new();
        public List<Company> Suppliers { get; set; } = new();
        public List<QuotationDropdownDto> Quotations { get; set; } = new();
        public List<PurchaseOrder> PurchaseOrders { get; set; } = new();
        public List<DODropdownDto> DeliveryOrders { get; set; } = new();
        public List<Project> Projects { get; set; } = new();
        public List<SalesOrderDropdownDto> SalesOrders { get; set; }
    }

    public class ConvertPOToInvoiceRequest
    {
        public Guid PoId { get; set; }
        public decimal Amount { get; set; }
    }

    public class PurchaseOrderDropdownDto
    {
        public List<QuotationDropdownDto> Quotations { get; set; } = new();

        public List<ProjectDropdownItem> Projects { get; set; } = new();
        public List<CompanyDropdownItem> Companies { get; set; } = new();
        public List<CompanyDropdownItem> Suppliers { get; set; } = new();
        public List<CompanyDropdownItem> Clients { get; set; } = new();
        public List<UserDto> Users { get; set; } = new();

    }

}

File: frontend/src/app/models/PurchaseOrder.ts
import { BaseModel } from './BaseModel';
import { CompanyDto } from './Company';
import { InvoiceDto } from './Invoice';
import { ProjectDto } from './Project';
import { QuotationDto } from './Quotation';
import { SalesOrderItem } from './SalesOrder';
import { UserDto } from './User';
import { SalesOrderDto } from './SalesOrder';

export interface PurchaseOrderDto extends BaseModel {
  purchaseOrderNo: string;
  fromCompanyId: string;
  fromCompany: CompanyDto;
  poDate: Date;
  poReceivedDate: Date;
  supplierId: string;
  supplier: CompanyDto;
  paymentTerms: string;
  projectId: string;
  project: ProjectDto;
  quotationId?: string;
  quotation?: QuotationDto;
  salesOrderId: string;
  salesOrder: SalesOrderDto;
  totalQuantity: number;
  gross: number;
  discount: number;
  totalAmount: number;
  remarks: string;
  notes: string;
  poClientNo: string;
  soClientNo: string;
  status: string;
  termsAndConditions: string;
  bankDetails: string;
  attachment: string;
  purchaseOrderStatusHistories: PurchaseOrderStatusHistory[];
  purchaseOrderItems: PurchaseOrderItem[];
  invoices: InvoiceDto[];
}

export interface PurchaseOrderStatusHistory extends BaseModel {
  purchaseOrderId: string;
  purchaseOrder: PurchaseOrderDto;
  status: string;
  actionAt: Date;
  actionUserId: string;
  actionUser: UserDto;
  remarks: string;
  signatureImage: string;
}

export interface PurchaseOrderItem extends BaseModel {
  purchaseOrderId: string;
  purchaseOrder: PurchaseOrderDto;
  salesOrderItemId: string;
  salesOrderItem: SalesOrderItem;
  item: string;
  description: string;
  quantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  totalPrice: number;
  receivedQuantity: number;
}

export interface POItemBase {
  salesOrderItemId: string;
  item: string;
  description: string;
  quantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  totalPrice: number;
  receivedQuantity: number;
}

export interface POItemRequest extends POItemBase {}

export interface UpdatePOItemRequest extends POItemBase {
  id: string;
}

export interface CreatePORequest {
  purchaseOrderNo: string;
  fromCompanyId: string;
  poDate: Date;
  poReceivedDate: Date;
  supplierId?: string;
  paymentTerms: string;
  quotationId: string;
  salesOrderId: string;
  projectId: string;
  gross: number;
  discount: number;
  totalAmount: number;
  notes: string;
  remarks: string;
  termsAndConditions: string;
  bankDetails: string;
  totalQuantity: number;
  attachment?: string;
  purchaseOrderItems: POItemBase[];
}

export interface UpdatePORequest extends CreatePORequest {
  id: string;
  purchaseOrderItems: UpdatePOItemRequest[];
}

export interface UpdatePOStatusRequest {
  id: string;
  status: string;
  remarks: string;
  signatureImage: string;
}

export interface PurchaseOrderDropdownDto {
  clients: CompanyDto[];
  suppliers: CompanyDto[];
  quotations: QuotationDto[];
}
